features/network_failures.feature

```diff
@@ -0,0 +1,109 @@
+    Then stderr contains substring:
+      """
+      Error connecting to /v1/context/machines/token: 503 {"error": "Bad Gateway"}
+      """
+    Then the machine is unattached
+
+    Examples: ubuntu release
+      | release | machine_type  |
+      | xenial  | lxd-container |
+      | noble   | lxd-container |
+
+  Scenario Outline: Network errors for attaching contract token are handled gracefully
+    # This test simulates network failure by disabling internet connection
+    # and then trying to attach contract token
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I disable any internet connection on the machine
+    And I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Failed to attach machine. See https://ubuntu.com/pro/dashboard
+      """
+    Then the machine is unattached
+
+    Examples: ubuntu release
+      | release | machine_type  |
+      | xenial  | lxd-container |
+      | noble   | lxd-container |
+
+  Scenario Outline: Network errors for enabling Realtime kernel are handled gracefully
+    # This test simulates network failure by disabling internet connection
+    # and then trying to enable realtime-kernel
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I attach `contract_token` with sudo and options `--no-auto-enable`
+    Then the machine is attached
+    Then I verify that `realtime-kernel` is disabled
+    When I disable any internet connection on the machine
+    And I verify that running `pro enable realtime-kernel` `with sudo` and stdin `y\ny` exits `1`
+    Then stderr contains substring:
+      """
+      Failed to connect to https://contracts.canonical.com/v1/contracts/
+      """
+    Then I verify that `realtime-kernel` is disabled
+
+    # Realtime kernel is not supported on LXD containers so we must use a VM
+    Examples: ubuntu release
+      | release | machine_type |
+      | xenial  | lxd-vm       |
+      | noble   | lxd-vm       |
+
+  Scenario Outline: Network errors for enabling Livepatch are handled gracefully
+    # This test simulates network failure by disabling internet connection
+    # and then trying to enable livepatch
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I attach `contract_token` with sudo and options `--no-auto-enable`
+    Then the machine is attached
+    Then I verify that `livepatch` is disabled
+    When I disable any internet connection on the machine
+    Then I verify that running `pro enable livepatch` `with sudo` exits `1`
+    Then stderr contains substring:
+      """
+      Failed to connect to https://contracts.canonical.com/v1/contracts/
+      """
+    Then I verify that `livepatch` is disabled
+
+    Examples: ubuntu release
+      | release | machine_type  |
+      | xenial  | lxd-container |
+      | noble   | lxd-container |
```

Comment:
See how this test is similar to the upper one, except for the service to be disabled?
You can create another column in the table below with "service to disable" so it looks like this:

```suggestion
    Examples: ubuntu release
      | release | machine_type  | service         |
      | xenial  | lxd-container | livepatch       |
      | xenial  | lxd-container | realtime-kernel |
      | noble   | lxd-container | livepatch       |
      | noble   | lxd-container | realtime-kernel |
```

Feel free to ignore this though - if you think this is good as is, keep it.

Reply:
At first glance, I was initially onboard because i love reducing duplicate code where reasonable, but then I noticed the `and stdin 'y\ny'` that only exists in the realtime kernel tests and not the livepatch tests.

Reply:
good catch forget what I said

Reply:
Instead, would using `--assume-yes` on the enable call allow you to avoid the `and stdin` so that they can be the same for livepatch and realtime?

Reply:
Oh you can do that??? That would be way better!